units/src/block.rs

```diff
@@ -0,0 +1,266 @@
+// SPDX-License-Identifier: CC0-1.0
+
+//! Block height and interval types.
+//!
+//! These types are thin wrappers around `u32`, no invariants implemented or implied.
+//!
+//! These are general types for abstracting over block heights, they are not designed to use with
+//! lock times. If you are creating lock times you should be using the
+//! [`locktime::absolute::Height`] and [`locktime::relative::Height`] types.
+//!
+//! The difference between these types and the locktime types is that these types are thin wrappers
+//! whereas the locktime types contain more complex locktime specific abstractions.
+
+use core::{fmt, ops};
+
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
+#[cfg(doc)]
+use crate::locktime;
+use crate::locktime::{absolute, relative};
+
+/// The block height, zero denotes the genesis block.
+///
+/// This type is not meant for constructing height based timelocks, this is a general purpose block
+/// height abstraction. For locktimes please see [`locktime::absolute::Height`].
+///
+/// This is a thin wrapper around a `u32` that may take on all values of a `u32`.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+// Public to try and make it really clear that there are no invariants.
+pub struct BlockHeight(pub u32);
+
+impl BlockHeight {
+    /// Block height 0, the genesis block.
+    pub const ZERO: Self = BlockHeight(0);
+
+    /// The minimum block height (0), the genesis block.
+    pub const MIN: Self = Self::ZERO;
+
+    /// The maximum block height.
+    pub const MAX: Self = BlockHeight(u32::MAX);
+
+    /// Creates a block height from a `u32`.
+    // Because From<u32> is not const.
+    pub const fn from_u32(inner: u32) -> Self { Self(inner) }
+
+    /// Returns block height as a `u32`.
+    // Because type inference doesn't always work using `Into`.
+    pub const fn to_u32(&self) -> u32 { self.0 }
+}
+
+impl fmt::Display for BlockHeight {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result { fmt::Display::fmt(&self.0, f) }
+}
+
+crate::impl_parse_str_from_int_infallible!(BlockHeight, u32, from);
+
+impl From<u32> for BlockHeight {
+    fn from(inner: u32) -> Self { Self::from_u32(inner) }
+}
```

Comment:
No, I don't like this just because it's a "thin" wrapper. My reasoning is to discourage less-readable `fn_taking_height(unitless_variable.into())`. So I'd much prefer to remove this.

Reply:
In your example the function itself has `height` in its name so it is clear what is being converted to. And if `unitless_variable` is defined immediately above, e.g. as an argument to a function, then it will be clearly a `u32`.

If you find this unreadable it's fine not to use it. But I think we should have it.

Reply:
I only added the name for my argument to be obvious but looks like it made it less so. :D

I think the best approach would be to have a lint for that but I don't see it happening.